x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -172,4 +210,49 @@ private void removeSessionForShard(String sessionUUID, IndexShard indexShard) {
             }
         }
     }
+
+    private static class RefCountedCloseable<T extends Closeable> extends AbstractRefCounted {
+
+        private final String name;
+        private final T object;
+
+        private RefCountedCloseable(String name, T object) {
+            super(name);
+            this.name = name;
+            this.object = object;
+        }
+
+        @Override
+        protected void closeInternal() {
+            IOUtils.closeWhileHandlingException(object);
+        }
+    }
+
+    public static class Reader implements Closeable {
+
+        private final RefCountedCloseable<Engine.IndexCommitRef> session;
+        private final RefCountedCloseable<IndexInput> input;
+
+        private Reader(RefCountedCloseable<Engine.IndexCommitRef> session, RefCountedCloseable<IndexInput> input) {
+            this.session = session;
+            this.input = input;
+            session.incRef();
+            input.incRef();
+        }
+
+        public void readFileBytes(byte[] chunk, long offset, int length) throws IOException {
+            synchronized (input.object) {
+                try (IndexInput in = input.object.clone()) {
```

Comment:
cloning IndexInput can be very very expensive since depending on the impl we clone a big buffer. Why is this necessary here? Can't se simplify concurrency here that only one consumer reads from the stream?